rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,664 @@
+    _files = desc.get_files();
+    _seed = desc.seed();
+    ret = subfolder_reading();
+    _file_headers.resize(_file_names.size());
+    // shuffle dataset if set
+    if (ret == Reader::Status::OK && _shuffle) {
+        std::mt19937 rng(_seed);
+        std::shuffle(_file_names.begin() + _shard_start_idx_vector[_shard_id],
+                     _file_names.begin() + _shard_end_idx_vector[_shard_id], rng);
+    }
+    return ret;
+}
+
+void NumpyDataReader::incremenet_read_ptr() {
+    _read_counter++;
+    increment_curr_file_idx(_file_names.size());
+}
+
+size_t NumpyDataReader::open() {
+    auto file_path = _file_names[_curr_file_idx];       // Get current file name
+    _curr_file_header = _file_headers[_curr_file_idx];  // Get current file header
+    incremenet_read_ptr();
+    _last_file_path = _last_id = file_path;
+    auto last_slash_idx = _last_id.find_last_of("\\/");
+    if (std::string::npos != last_slash_idx) {
+        _last_id.erase(0, last_slash_idx + 1);
+    }
+
+    _header_parsing_failed = false;
+    auto ret = get_header_from_cache(file_path, _curr_file_header);
+    if (!ret) {
+        parse_header(_curr_file_header, file_path);
+        if(_header_parsing_failed) {
+            ERR("Numpy header parsing failed");
+            return 0;
+        }
+        update_header_cache(file_path, _curr_file_header);
+    } else {
+        _current_file_ptr = std::fopen(file_path.c_str(), "rb");
+        if (_current_file_ptr == nullptr) {
+            ERR("Could not open file " + file_path + ": " + std::strerror(errno));
+            return 0;
+        }
+    }
+    fseek(_current_file_ptr, 0, SEEK_SET);  // Take the file pointer back to the start
+    return _curr_file_header.nbytes();
+}
+
+bool NumpyDataReader::get_header_from_cache(const std::string& file_name, NumpyHeaderData& header) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    auto it = _header_cache.find(file_name);
+    if (it == _header_cache.end()) {
+        return false;
+    } else {
+        header = it->second;
+        return true;
+    }
+}
+
+void NumpyDataReader::update_header_cache(const std::string& file_name, const NumpyHeaderData& value) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    _header_cache[file_name] = value;
+}
+
+const RocalTensorDataType NumpyDataReader::get_dtype(const std::string& format) {
+    if (format == "u1") return RocalTensorDataType::UINT8;
+    if (format == "u2") THROW("uint16_t dtype not supported in rocAL");
+    if (format == "u4") return RocalTensorDataType::UINT32;
+    if (format == "u8") THROW("uint64_t dtype not supported in rocAL");
+    if (format == "i1") return RocalTensorDataType::INT8;
+    if (format == "i2") return RocalTensorDataType::INT16;
+    if (format == "i4") return RocalTensorDataType::INT32;
+    if (format == "i8") THROW("int64_t dtype not supported in rocAL");
+    if (format == "f2")
+#if defined(AMD_FP16_SUPPORT)
```

Comment:
why do we need to check for this flag. It should be supported by default in all GPUs

Reply:
If half library is found in CMake, this flag is set. We are checking for this flag in tensor.h and then only enable vx_float16 dtype if the flag is set

Reply:
Cannot use switch case for this function since strings are not supported. maps cannot be used since we need to throw an error for the unsupported types.